README.md
# Ataraxy

[![Build Status](https://travis-ci.org/weavejester/ataraxy.svg?branch=master)](https://travis-ci.org/weavejester/ataraxy)

A data-driven routing and destructuring library for [Ring][]. This
library is still being developed, so some functionality may change
before we hit version 1.0.0.

[ring]: https://github.com/ring-clojure/ring

## Rationale

There are several data-driven routing libraries for Ring, such as
[bidi][], [Silk][] and [gudu][]. Ataraxy differs from them because
it not only seeks to match a route, it also destructures the
incoming request.

In this sense it is similar to [Compojure][], in that the idea is to
remove extraneous information. However, while Compojure is designed to
use chains of functions, Ataraxy defines its functionality through a
declarative data structure.

[bidi]: https://github.com/juxt/bidi
[silk]: https://github.com/DomKM/silk
[gudu]: https://github.com/thatismatt/gudu
[compojure]: https://github.com/weavejester/compojure


## Example

```clojure
{["/api" {uid :identity}]
 {"/products"
   {[:get]                [:products/list uid]
    [:get "/" pid]        [:products/get uid ^uuid pid]
    [:get "/search" #{q}] [:products/search uid q]
    [:post {body :body}]  [:products/new uid body]}}}
```


## Installation

Add the following dependency to your `project.clj` file:

    [ataraxy "0.4.3"]


## Routing

Ataraxy uses a data structure to tell it how to route and destructure
requests. See the following section on [syntax](#syntax) for details.

```clojure
(def routes '{"/foo" [:foo]})
```

We can match a request map to a result with `matches`:

```clojure
(require '[ataraxy.core :as ataraxy])

(ataraxy/matches routes {:uri "/foo"})
=> [:foo]
```

If Ataraxy cannot correctly match any route, then an error result from
the `ataraxy.error` namespace is returned. For example:

```clojure
(ataraxy/matches routes {:uri "/bar"})
=> [:ataraxy.error/unmatched-path]
```

See the [errors](#errors) section for more details.

For performance, we can also pre-compile the routing data:

```clojure
(def compiled-routes (ataraxy/compile routes))
```

The resulting object can be used in `matches` in the same way as the
raw data structure:

```clojure
(ataraxy/matches compiled-routes {:uri "/foo"})
=> [:foo]
```


## Handlers

Once we have our routes, it's likely we want to turn them into a Ring
handler function. Ataraxy has a function called `handler` for this
purpose:

```clojure
(defn foo [request]
  {:status 200, :headers {}, :body "Foo"})

(def handler
  (ataraxy/handler
   {:routes   routes
    :handlers {:foo foo}}))
```

This function takes a map with four keys:

* `:routes`     - the routes to match
* `:handlers`   - a map of result keys to Ring handlers
* `:middleware` - a map of metadata keys to Ring middleware (optional)
* `:coercers`   - a map of symbols to coercer functions (optional)

The handler function is chosen by the key of the result. Two keys are
added to the request map passed to the handler:

* `:ataraxy/result` - contains the matched result
* `:route-params`   - a map of parameters matched in the path
                      (included for compatibility)

The handler can also return a result vector instead of a request
map. Each vector that is returned is checked against the handler map,
until eventually a Ring response map is returned.

The `ataraxy.response` namespace defines a number of responses on
the default handler, allowing for code like this:

```clojure
(require '[ataraxy.response :as response])

(defn hello [{[_ name] :ataraxy/result}]
  [::response/ok (str "Hello " name)])

(def handler
  (ataraxy/handler
    {:routes   '{[:get "/hello/" name] [:hello name]}
     :handlers {:hello hello}}))
```

The default handler is set to `ataraxy.handler/default`, but can be
changed by adding a handler to the `:default` key of the handler map.

Middleware is chosen based on the metadata that is applied to the
result or to the containing routing table. For example:

```clojure
(defn wrap-example [handler value]
  (fn [request]
    (let [response (handler request)]
      (assoc-in response [:header "X-Example"] value))))

(def handler
  (ataraxy/handler
   {:routes     {"/foo" ^:example [:foo]}
    :handlers   {:foo foo}
    :middleware {:example #(wrap-example % "test")}}))
```

This would add an `X-Example` header to the response of the
handler. We can also pass an argument to the handler by setting the
`:example` metadata key to something other than `true`:

```clojure
(def handler
  (ataraxy/handler
   {:routes     {"/foo" ^{:example "test"} [:foo]}
    :handlers   {:foo foo}
    :middleware {:example wrap-example}}))
```

Custom coercers can be added to the handler by specifying the
`:coercers` option. This is described in more detail in
the [coercers](#coercers) section.


## Syntax

Ataraxy generates routes from a **routing table**, which is a Clojure
map, or a list of alternating keys and values.

The keys of the table are **routes**, and the data type used defines a
way of matching and destructuring a request.

The values are either **results** or nested tables.

Here's a semi-formal definition of the syntax:

```
table  = {<route result>+} | (<route result>+)
route  = keyword | string | symbol | set | map | [route+]
result = table | [keyword symbol*]
```

### Results

Results are always vectors, beginning with a keyword, followed by zero
or more symbols. For example:

```clojure
[:foo id]
```

Results are paired with routes:

```clojure
{["/foo/" id] [:foo id]}
```

The symbols in the route are passed into the result.

The symbols in the result may be tagged with a type they should be
coerced into. For example:

```clojure
[:foo ^int id]
```

See the [coercers](#coercers) section for more detail.

### Keyword routes

A keyword will match the request method. For example:

```clojure
{:get [:foo]})
```

This route will match any request with the GET method.

### String routes

A string will match the `:path-info` or `:uri` key on the request. For
example:

```clojure
{"/foo" [:foo]
 "/bar" [:bar]}
```

This example will match the URIs "/foo" and "/bar".

### Symbol routes

Like strings, symbols match against the `:path-info` or `:uri` key on
the request. Unlike strings, they match on a regular expression, and
bind the string matched by the regular expression to the symbol.

By default the regex used is `[^/]+`. In other words, any character
except a forward slash. The regex can be changed by adding a `:re` key
to the symbol's metadata. For example:

```clojure
{^{:re #"/d.g"} w [:word w]}
```

This will match URIs like "/dog", "/dig" and "/dug", and add the
matched word to the result.

### Set routes

A set of symbols will match URL-encoded parameters of the same name.
For example:

```clojure
{#{q} [:query q]}
```

This will match any request with `q` as a parameter. For example,
"/search?q=foo".

In order to match query parameters, the request map needs a
`:query-params` key, which is supplied by the [wrap-params][]
middleware in Ring.

By default, the parameters must be set for the route to match. If you
want the parameters to be optional, you can prefix them with a "?".

```clojure
{#{?q} [:query ?q]}
```

This works the same as the previous example, except that the route
still matches if `q` is `nil`.

[wrap-params]: https://ring-clojure.github.io/ring/ring.middleware.params.html#var-wrap-params

### Map routes

A map will destructure the request. Any destructured symbol must not
be `nil` for the route to match. For example:

```clojure
{{{:keys [user]} :session} [:user user]}
```

This route will match any request map with a `:user` key in the
session.

As with set routes, symbols prefixed with a "?" are considered
optional and may be `nil`.


### Vector routes

A vector combines the behavior of multiple routing rules. For example:

```clojure
{[:get "/foo"] [:foo]}
```

This will match both the request method and the URI.

Strings and symbols will be combined in order, to allow complex paths
to be matched. For example:

```clojure
{[:get "/user/" name "/info"] [:get-user-info name]}
```

This will match URIs like "/user/alice/info" and pass the name "alice"
to the result.

### Nested tables

Nesting routing tables is an alternative way of combining routes.
Instead of a result vector, a map or list may be specified. For
example:

```clojure
{"/foo"
 {"/bar" [:foobar]
  "/baz" [:foobaz]}})
```

This will match the URIs "/foo/bar" and "/foo/baz".

You can also use nesting and vectors together:

```clojure
{["/user/" name]
 {:get [:get-user name]
  :put [:put-user name]}}
```


## Errors

When something goes wrong, Ataraxy returns one of the following error
results:

* `:ataraxy.error/unmatched-path`
* `:ataraxy.error/unmatched-method`
* `:ataraxy.error/missing-params`
* `:ataraxy.error/missing-destruct`
* `:ataraxy.error/failed-coercions`
* `:ataraxy.error/failed-spec`

If you're using the `ataraxy.core/handler` function, these are
automatically converted into appropriate Ring response maps. However,
it's generally worth customizing the error responses to the needs of
your application.


## Coercers

Coercers are functions that turn a string into a custom type. Any
symbol in the result can be tagged with a symbol associated with a
coercer function.

For example, it's common to want to change a parameter from a string
into an int:

```clojure
{[:get "/foo/" id] [:foo ^int id]}
```

The `int` and `uuid` coercers are included by default. We can easily
add our own, however:

```clojure
(defn ->float [s]
  (try (Double/parseDouble s) (catch NumberFormatException _)))

(def compiled-routes
  (ataraxy/compile
   '{[:get "/foo/" id] [:foo ^float id]}
   {'float ->float}))
```

And similarly to handlers:

```clojure
(def handler
  (ataraxy/handler
   {:routes   '{[:get "/foo/" id] [:foo ^float id]}
    :coercers {'float ->float}}))
```


## Specs

Results are validated via the `:ataraxy/result` spec. This is a
multi-spec that dispatches off the key, and can be assigned behavior
through the `result-spec` multimethod.

For example:

```clojure
(require '[clojure.spec.alpha :as s])

(defmethod ataraxy/result-spec ::foo [_]
  (s/cat :key any? :id nat-int?))
```

This ensures that any result with `::foo` as the key must have exactly
two elements, with the second being a natural number.

If a spec fails, then a `:ataraxy.error/failed-spec` result is
returned, which if left alone resolves to a 400 "Bad Request"
response in the handler.


## License

Copyright © 2022 James Reeves

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
## 0.4.3 (2022-04-20)

* Added ::unprocessable-entity response (#22)

## 0.4.2 (2018-09-28)

* Fixed issue with coercions (#18)

## 0.4.1 (2018-09-03)

* Fixed broken routing table spec in Clojure 1.10 (#14)

## 0.4.0 (2017-06-26)

* Added extensible spec validation to results


test/ataraxy/core_test.clj
(ns ataraxy.core-test
  (:require [clojure.spec.alpha :as s]
            [clojure.test :refer :all]
            [ataraxy.core :as ataraxy]
            [ataraxy.error :as err]
            [ataraxy.response :as response]
            [clojure.java.io :as io]))

(defmethod ataraxy/result-spec ::foo [_]
  (s/cat :key keyword? :id (s/and int? #(> % 10))))

(deftest test-valid?
  (are [x y] (= (ataraxy/valid? x) y)
    '{"/foo" [:bar]}        true
    '{}                     true
    '{"/foo" :bar}          false
    '{"/foo" (:bar)}        false
    '{"/x" [:x], "/y" [:x]} false))

(deftest test-matches
  (testing "string routes"
    (let [routes '{"/" [:idx] "/foo" [:foo], "/bar" [:bar]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/"}          [:idx]
        {:uri "/foo"}       [:foo]
        {:path-info "/bar"} [:bar]
        {:uri "/baz"}       [::err/unmatched-path])))

  (testing "list routing tables"
    (let [routes '("/foo" [:foo], "/bar" [:bar])]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/foo"}       [:foo]
        {:path-info "/bar"} [:bar]
        {:uri "/baz"}       [::err/unmatched-path])))

  (testing "symbol routes"
    (let [routes '{x [:foo x]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "foo"}       [:foo "foo"]
        {:path-info "bar"} [:foo "bar"])))

  (testing "custom regexes"
    (let [routes '{^{:re #"\d\d"} x   [:foo x]
                   ^{:re #"\d\d\d"} y [:bar y]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "10"}     [:foo "10"]
        {:uri "1"}      [::err/unmatched-path]
        {:uri "bar"}    [::err/unmatched-path]
        {:uri "10/bar"} [::err/unmatched-path]
        {:uri "200"}    [:bar "200"]
        {:uri "1"}      [::err/unmatched-path])))

  (testing "keyword routes"
    (let [routes '{:get [:read], :put [:write]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :get}    [:read]
        {:request-method :put}    [:write]
        {:request-method :delete} [::err/unmatched-method])))

  (testing "set routes"
    (let [routes '{#{x} [:foo x], #{y} [:bar y], #{z w} [:baz z w]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/" :query-params {"x" "quz"}}       [:foo "quz"]
        {:uri "/" :query-params {"y" "quz"}}       [:bar "quz"]
        {:uri "/" :query-params {"z" "a" "w" "b"}} [:baz "a" "b"]
        {:uri "/" :form-params {"x" "fp"}}         [:foo "fp"]
        {:uri "/" :multipart-params {"x" "mp"}}    [:foo "mp"]
        {:uri "/" :query-params {"z" "a"}}         [::err/missing-params '#{x}]
        {:uri "/"}                                 [::err/missing-params '#{x}])))

  (testing "map routes"
    (let [routes '{{{p :p} :params} [:p p], {{:keys [q]} :params} [:q q]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:params {:p "page"}}    [:p "page"]
        {:params {:q "query"}}   [:q "query"]
        {:params {:z "invalid"}} [::err/missing-destruct '#{p}])))

  (testing "optional bindings"
    (let [routes '{["/p" #{?p}] [:p ?p]
                   ["/q" {{?q "q"} :query-params}] [:q ?q]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/p", :query-params {"p" "page"}}  [:p "page"]
        {:uri "/p", :query-params {"q" "query"}} [:p nil]
        {:uri "/q", :query-params {"q" "query"}} [:q "query"]
        {:uri "/q", :query-params {"p" "page"}}  [:q nil]
        {:uri "/z", :query-params {"p" "page"}}  [::err/unmatched-path]
        {:uri "/z", :query-params {"q" "query"}} [::err/unmatched-path])))

  (testing "compiled routes"
    (let [routes (ataraxy/compile '{"/foo" [:foo]})]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/foo"} [:foo]
        {:uri "/bar"} [::err/unmatched-path])))

  (testing "vector routes"
    (let [routes '{["/foo/" foo]          [:foo foo]
                   [:get "/bar"]          [:bar]
                   ["/baz" #{baz}]        [:baz baz]
                   [:get "/x/" x "/y/" y] [:xy x y]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:uri "/foo/10"}                         [:foo "10"]
        {:request-method :get, :uri "/bar"}      [:bar]
        {:uri "/baz", :query-params {"baz" "2"}} [:baz "2"]
        {:request-method :get, :uri "/x/8/y/3a"} [:xy "8" "3a"]
        {:uri "/foo"}                            [::err/unmatched-path]
        {:request-method :put, :uri "/bar"}      [::err/unmatched-method]
        {:request-method :get, :uri "/x/44/y/"}  [::err/unmatched-path])))

  (testing "nested routes"
    (let [routes '{"/foo" {:get [:foo], ["/" id] {:get [:bar id]}}}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :get, :uri "/foo"}    [:foo]
        {:request-method :get, :uri "/foo/10"} [:bar "10"]
        {:request-method :put, :uri "/foo"}    [::err/unmatched-method])))

  (testing "coercions"
    (let [routes '{["/foo/" id]  [:foo ^int id]
                   ["/bar/" id]  [:bar ^uuid id]
                   ["/baz" #{?p}] [:baz ^int ?p]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :get, :uri "/foo/10"} [:foo 10]
        {:request-method :get, :uri "/foo/xx"} [::err/failed-coercions '#{id}]

        {:request-method :get, :uri "/bar/5eae6ec3-f4eb-477e-b255-a6dd686bb697"}
        [:bar #uuid "5eae6ec3-f4eb-477e-b255-a6dd686bb697"]

        {:request-method :get, :uri "/bar/5eae6ec3"}
        [::err/failed-coercions '#{id}]

        {:request-method :get, :uri "/baz", :query-params {"p" "10"}} [:baz 10]
        {:request-method :get, :uri "/baz", :query-params {}}         [:baz nil])))

  (testing "custom coercers"
    (let [routes (ataraxy/compile
                  '{["/foo/" id] [:foo ^second id]}
                  {'second second})]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :get, :uri "/foo/bar"} [:foo \a]
        {:request-method :get, :uri "/foo/b"}   [::err/failed-coercions '#{id}])))

  (testing "custom coercers without matching classes"
    (let [routes (ataraxy/compile
                  '{["/foo/" a "/" b] [:foo ^xyz a ^xyz b]}
                  {'xyz #(str % ".")})]
      (is (= (ataraxy/matches routes {:request-method :get, :uri "/foo/bar/baz"})
             [:foo "bar." "baz."]))))

  (testing "error results"
    (let [routes '{[:get "/foo/" id #{page}] [:foo id ^int page]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :put, :uri "/foo"}    [::err/unmatched-path]
        {:request-method :put, :uri "/foo/10"} [::err/unmatched-method]
        {:request-method :get, :uri "/foo/10"} [::err/missing-params '#{page}]
        {:request-method :get, :uri "/foo/10"
         :query-params {"page" "x"}}           [::err/failed-coercions '#{page}])))

  (testing "specs"
    (let [routes '{[:get "/foo/" id] [::foo ^int id]}]
      (are [req res] (= (ataraxy/matches routes req) res)
        {:request-method :get, :uri "/foo/20"} [::foo 20]
        {:request-method :get, :uri "/foo/10"}
        [::err/failed-spec
         '#:clojure.spec.alpha{:problems
                               [{:path [:ataraxy.core-test/foo :id],
                                 :pred
                                 (clojure.core/fn [%] (clojure.core/> % 10)),
                                 :val 10,
                                 :via [:ataraxy/result],
                                 :in [1]}],
                               :spec :ataraxy/result,
                               :value [:ataraxy.core-test/foo 10]}]))))

(deftest test-handler
  (testing "synchronous handler"
    (let [handler (ataraxy/handler
                   {:routes
                    '{[:get "/foo"]     [:foo]
                      [:get "/bar/" id] [:bar id]}
                    :handlers
                    {:foo
                     (constantly {:status 200, :headers {}, :body "foo"})
                     :bar
                     (fn [{[_ id] :ataraxy/result}]
                       {:status 200, :headers {}, :body (str "bar" id)})}})]
      (is (= (handler {:request-method :get, :uri "/foo"})
             {:status 200, :headers {}, :body "foo"}))
      (is (= (handler {:request-method :get, :uri "/bar/baz"})
             {:status 200, :headers {}, :body "barbaz"}))
      (is (= (handler {:request-method :get, :uri "/baz"})
             {:status  404
              :headers {"Content-Type" "text/html; charset=UTF-8"}
              :body    "Not Found"}))
      (is (= (handler {:request-method :put, :uri "/foo"})
             {:status  405
              :headers {"Content-Type" "text/html; charset=UTF-8"}
              :body    "Method Not Allowed"}))))

  (testing "asynchronous handler"
    (let [handler (ataraxy/handler
                   {:routes
                    '{[:get "/foo"]     [:foo]
                      [:get "/bar/" id] [:bar id]}
                    :handlers
                    {:foo
                     (fn [request respond raise]
                       (respond {:status 200, :headers {}, :body "foo"}))
                     :bar
                     (fn [{[_ id] :ataraxy/result} respond raise]
                       (respond {:status 200, :headers {}, :body (str "bar" id)}))}})]
      (let [respond (promise), raise (promise)]
        (handler {:request-method :get, :uri "/foo"} respond raise)
        (is (= @respond {:status 200, :headers {}, :body "foo"})))
      (let [respond (promise), raise (promise)]
        (handler {:request-method :get, :uri "/bar/baz"} respond raise)
        (is (= @respond {:status 200, :headers {}, :body "barbaz"})))
      (let [respond (promise), raise (promise)]
        (handler {:request-method :get, :uri "/baz"} respond raise)
        (is (= @respond {:status  404
                         :headers {"Content-Type" "text/html; charset=UTF-8"}
                         :body    "Not Found"})))
      (let [respond (promise), raise (promise)]
        (handler {:request-method :put, :uri "/foo"} respond raise)
        (is (= @respond {:status  405
                         :headers {"Content-Type" "text/html; charset=UTF-8"}
                         :body    "Method Not Allowed"})))))

  (testing "middleware"
    (let [handler (ataraxy/handler
                   {:routes
                    '{"/foo" {:get ^:baz [:foo]}
                      "/bar" ^{:quz 9} {["/" id] {:get [:bar id]}}}
                    :handlers
                    {:foo
                     (constantly {:status 200, :headers {}, :body "foo"})
                     :bar
                     (fn [{[_ id] :ataraxy/result}]
                       {:status 200, :headers {}, :body (str "bar" id)})}
                    :middleware
                    {:baz
                     (fn [handler]
                       #(assoc-in (handler %) [:headers "X-Middle"] "baz"))
                     :quz
                     (fn [handler x]
                       #(assoc-in (handler %) [:headers "X-Middle"] (str "quz" x)))}})]
      (is (= (handler {:request-method :get, :uri "/foo"})
             {:status 200, :headers {"X-Middle" "baz"}, :body "foo"}))
      (is (= (handler {:request-method :get, :uri "/bar/10"})
             {:status 200, :headers {"X-Middle" "quz9"}, :body "bar10"}))
      (is (= (handler {:request-method :get, :uri "/baz"})
             {:status  404
              :headers {"Content-Type" "text/html; charset=UTF-8"}
              :body    "Not Found"}))))

  (testing "route parameters"
    (let [handler (ataraxy/handler
                   {:routes
                    '{[:get "/user/" uid "/post/" pid] [:user-post uid pid]}
                    :handlers
                    {:user-post
                     (fn [{params :route-params}]
                       {:status 200, :headers {}, :body params})
                     ::err/unmatched-method
                     (fn [{params :route-params}]
                       {:status 405, :headers {}, :body params})}})]
      (is (= (handler {:request-method :get, :uri "/user/alice/post/5"})
             {:status 200, :headers {}, :body {:uid "alice" :pid "5"}}))
      (is (= (handler {:request-method :put, :uri "/user/alice/post/5"})
             {:status 405, :headers {}, :body {:uid "alice" :pid "5"}}))))

  (testing "variant responses"
    (let [handler (ataraxy/handler
                   {:routes   '{[:get "/foo/" id] [:foo id]}
                    :handlers {:foo
                               (fn [{[_ id] :ataraxy/result}]
                                 [::response/ok (str "id=" id)])}})]
      (is (= (handler {:request-method :get, :uri "/foo/bar"})
             {:status  200
              :headers {"Content-Type" "text/html; charset=UTF-8"}
              :body    "id=bar"}))))

  (testing "resource responses"
    (let [resource (io/resource "ataraxy/foo.txt")
          handler  (ataraxy/handler
                     {:routes   '{[:get "/foo"] [:foo]}
                      :handlers {:foo (fn [_] [::response/ok resource])}})
          response (handler {:request-method :get, :uri "/foo"})]
      (is (= (:status response) 200))
      (is (= (get-in response [:headers "Content-Type"]) "text/plain"))
      (is (instance? java.io.File (:body response)))
      (is (= (slurp (:body response)) (slurp resource)))))

  (testing "file responses"
    (let [file     (io/file "test/ataraxy/foo.txt")
          handler  (ataraxy/handler
                     {:routes   '{[:get "/foo"] [:foo]}
                      :handlers {:foo (fn [_] [::response/ok file])}})
          response (handler {:request-method :get, :uri "/foo"})]
      (is (= (:status response) 200))
      (is (= (get-in response [:headers "Content-Type"]) "text/plain"))
      (is (= (:body response) file)))))


test/ataraxy/response_test.clj
(ns ataraxy.response-test
  (:require [ataraxy.handler :as handler]
            [ataraxy.response :as resp]
            [clojure.test :refer :all]))

(deftest test-content
  (is (= (handler/default {:ataraxy/result [::resp/ok "<h1>Hello World</h1>"]})
         {:status  200
          :headers {"Content-Type" "text/html; charset=UTF-8"}
          :body    "<h1>Hello World</h1>"})))

(deftest test-redirects
  (is (= (handler/default {:ataraxy/result [::resp/created "/new" "foo"]})
         {:status  201
          :headers {"Content-Type" "text/html; charset=UTF-8"
                    "Location"     "/new"}
          :body    "foo"}))
  (is (= (handler/default {:ataraxy/result [::resp/found "/redirect"]})
         {:status 302, :headers {"Location" "/redirect"}, :body nil}))
  (is (= (handler/default {:ataraxy/result [::resp/temporary-redirect "/redirect"]})
         {:status 307, :headers {"Location" "/redirect"}, :body nil})))



src/ataraxy/error.clj
(ns ataraxy.error
  "A namespace that contains Ataraxy's standard error results, and adds default
  handler methods for each one."
  (:require [ataraxy.handler :as handler]
            [ataraxy.response :as response]))

(def errors
  "A map of errors to their relative priority."
  {::unmatched-path   0
   ::unmatched-method 1
   ::missing-params   2
   ::missing-destruct 3
   ::failed-coercions 4
   ::failed-spec      5})

(defn error-result?
  "Return true if the result is an Ataraxy error result."
  [result]
  (contains? errors (first result)))

(defmethod handler/sync-default ::unmatched-path [_]
  [::response/not-found "Not Found"])

(defmethod handler/sync-default ::unmatched-method [_]
  [::response/method-not-allowed "Method Not Allowed"])

(defmethod handler/sync-default ::missing-params [_]
  [::response/bad-request "Bad Request"])

(defmethod handler/sync-default ::missing-destruct [_]
  [::response/bad-request "Bad Request"])

(defmethod handler/sync-default ::failed-coercions [_]
  [::response/bad-request "Bad Request"])

(defmethod handler/sync-default ::failed-spec [_]
  [::response/bad-request "Bad Request"])


src/ataraxy/handler.clj
(ns ataraxy.handler
  "A namespace that contains the default handler for Ataraxy routes.")

(defmulti sync-default
  "A synchronous Ring handler, defined as a multimethod that dispatches off the
  :ataraxy/result key in the requst map."
  (fn [request] (first (:ataraxy/result request))))

(defmulti async-default
  "An asynchronous Ring handler, defined as a multimethod that dispatches off
  the :ataraxy/result key in the requst map."
  (fn [request _ _] (first (:ataraxy/result request))))

(defmethod async-default :default [request respond raise]
  (respond (sync-default request)))

(defn default
  "The default handler for Ataraxy routes. Extend by using the sync-default and
  async-default multimethods."
  ([request] (sync-default request))
  ([request respond raise] (async-default request respond raise)))


src/ataraxy/response.clj
(ns ataraxy.response
  "A namespace that contains Ataraxy's standard responses, and adds default
  handler methods for each one."
  (:require [ataraxy.handler :as handler]
            [ring.util.mime-type :as mime]
            [ring.util.response :as resp]))

(defn- guess-content-type [response name]
  (if-let [mime-type (mime/ext-mime-type (str name))]
    (resp/content-type response mime-type)
    response))

(defprotocol ToResponse
  "A protocol used to transform a value into a response map."
  (->response [x]
    "Turn its argument into a valid Ring response map."))

(extend-protocol ToResponse
  String
  (->response [s]
    (-> (resp/response s) (resp/content-type "text/html; charset=UTF-8")))
  java.io.File
  (->response [f]
    (-> (resp/file-response (str f)) (guess-content-type f)))
  java.net.URL
  (->response [url]
    (-> (resp/url-response url) (guess-content-type url)))
  Object
  (->response [o]
    (resp/response o))
  nil
  (->response [_]
    (resp/response nil)))

(defmethod handler/sync-default ::continue [{[_] :ataraxy/result}]
  {:status 100, :headers {}, :body nil})

(defmethod handler/sync-default ::ok [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 200)))

(defmethod handler/sync-default ::created [{[_ url body] :ataraxy/result}]
  (-> (->response body) (resp/header "Location" url) (resp/status 201)))

(defmethod handler/sync-default ::accepted [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 202)))

(defmethod handler/sync-default ::no-content [{[_] :ataraxy/result}]
  {:status 204, :headers {}, :body nil})

(defmethod handler/sync-default ::moved-permanently [{[_ url] :ataraxy/result}]
  {:status 301, :headers {"Location" url}, :body nil})

(defmethod handler/sync-default ::found [{[_ url] :ataraxy/result}]
  {:status 302, :headers {"Location" url}, :body nil})

(defmethod handler/sync-default ::see-other [{[_ url] :ataraxy/result}]
  {:status 303, :headers {"Location" url}, :body nil})

(defmethod handler/sync-default ::not-modified [{[_] :ataraxy/result}]
  {:status 304, :headers {}, :body nil})

(defmethod handler/sync-default ::temporary-redirect [{[_ url] :ataraxy/result}]
  {:status 307, :headers {"Location" url}, :body nil})

(defmethod handler/sync-default ::permanent-redirect [{[_ url] :ataraxy/result}]
  {:status 308, :headers {"Location" url}, :body nil})

(defmethod handler/sync-default ::bad-request [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 400)))

(defmethod handler/sync-default ::unauthorized [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 401)))

(defmethod handler/sync-default ::forbidden [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 403)))

(defmethod handler/sync-default ::not-found [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 404)))

(defmethod handler/sync-default ::method-not-allowed [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 405)))

(defmethod handler/sync-default ::not-acceptable [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 406)))

(defmethod handler/sync-default ::request-timeout [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 408)))

(defmethod handler/sync-default ::conflict [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 409)))

(defmethod handler/sync-default ::gone [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 410)))

(defmethod handler/sync-default ::length-required [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 411)))

(defmethod handler/sync-default ::precondition-failed [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 412)))

(defmethod handler/sync-default ::payload-too-large [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 413)))

(defmethod handler/sync-default ::uri-too-long [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 414)))

(defmethod handler/sync-default ::unsupported-media-type [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 415)))

(defmethod handler/sync-default ::expectation-failed [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 417)))

(defmethod handler/sync-default ::unprocessable-entity [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 422)))

(defmethod handler/sync-default ::precondition-required [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 428)))

(defmethod handler/sync-default ::too-many-requests [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 429)))

(defmethod handler/sync-default ::request-header-fields-too-large [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 431)))

(defmethod handler/sync-default ::unavailable-for-legal-reasons [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 451)))

(defmethod handler/sync-default ::internal-server-error [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 500)))

(defmethod handler/sync-default ::not-implemented [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 501)))

(defmethod handler/sync-default ::service-unavailable [{[_ body] :ataraxy/result}]
  (-> (->response body) (resp/status 502)))
